### DISCUSSION TITLE
@UseRequestContext cause problem with loading nested references

### DISCUSSION THREAD

**Discussion by Morb0** [mikro-orm/mikro-orm#3415] [Q&A] [answered]
**Describe the bug**
All loads correct without `@UseRequestContext`.
But with `@UseRequestContext` when you try load nested reference (reference of loaded already reference) it return just `Ref`.

```typescript
// events.service.ts
@Injectable()
export class EventsService {
  constructor(private readonly orm: MikroORM) {}
  
  @OnEvent('user.created')
  @UseRequestContext()
  onCreated(event: UserCreatedEvent): Promise<void> {
    const user = await event.user.load();
    console.log(user); // UserEntity { ... }
    console.log(user.referrer); // Reference { entity: Ref<UserEntity> { id: '829f1619-9747-42fd-866d-aa33d3f35ed5' } }
    const referrer = await referral.referrer?.load();
    console.log(referrer); // Ref<UserEntity> { id: '829f1619-9747-42fd-866d-aa33d3f35ed5' }
    // `referrer` should be loaded `UserEntity`, not `Ref`
  }
}

// user.entity.ts
@Entity()
export class UserEntity {
  // ...
  @ManyToOne(() => UserEntity, { nullable: true, wrappedReference: true })
  referrer?: IdentifiedReference<UserEntity>;
}
```

**To Reproduce**
1. Apply `@UseRequestContext` decorator to method
2. Load reference prop from entity (next `ref1`)
3. Try to load reference prop of `ref1` entity
4. Receive just `Ref` object

**Expected behavior**
Instead of just `Ref` it should be `UserEntity`.

**Versions**

| Dependency | Version |
| - | - |
| node | 18.4.0 |
| typescript | 4.3.5 |
| mikro-orm | 5.3.1 |
| postgres | 5.3.1 |

**Comment by B4nan** [comment, marked as answer]
The fact that you use this decorator in a service is kind of a red flag for me. Maybe you nest the calls? You should always have only one `@UseRequestContext` and it needs to be on the top level - you should never call a method decorated with it from another method already decorated the same way.

The fact that you are getting an entity instance in the argument is another red flag. How was this entity created, how come it is managed (you lazy load relations on it, so it needs to be managed), by what EM? This only prooves you have multiple contexts. 

Btw the `Ref` is just how you see things logged, it is still a `UserEntity` class (you can try yourself via `instanceof`).

You should either load that user entity from inside the method, or merge it (via `em.merge()`).

**Reply by Morb0** [reply]
That method only handle events by `@OnEvent` decorator and not called anywhere else.

Entity creates on user account registration request and `@AfterCreate` orm hook emits `user.created` event, perhaps this is the problem, because there possible some race condition or loss entity manager.

`em.merge()` worked successfully, but getting entity in the method instead of passing it is better idea.

**Comment by Morb0** [comment]
@B4nan As I now know, hook `afterCreate` does not called after entity committed in DB, just after flush, so I can't really emit only ID because when I try to retrieve an entity by this ID in method it won't be in the database yet.

But I need to trigger events after the entity is created. What is the best way to do this with the ORM?

**Reply by B4nan** [reply]
> does not called after entity committed in DB

It is called after the EM flushes new entity to database. Yes, you need flush. But I dont understand what else are you asking for. There are no hooks for "insert queries" (e.g. via `em.nativeInsert()` or via QB) if that is what you mean. 

You can as well hook into transaction events, or maybe `afterFlush` is more in line with what you want?

Its all described here, I am sure you have seen that page already. Be more specific if you want help. 

https://mikro-orm.io/docs/events
